File: backend-kotlin/src/main/kotlin/com/alexn/socialpublish/db/PostsDatabase.kt
package com.alexn.socialpublish.db

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.jdbi.v3.core.Jdbi
import java.time.Instant

private val json = Json { ignoreUnknownKeys = true }

@Serializable
data class PostPayload(
    val content: String,
    val link: String? = null,
    val tags: List<String>? = null,
    val language: String? = null,
    val images: List<String>? = null
)

data class Post(
    val uuid: String,
    val targets: List<String>,
    val createdAt: Instant,
    val content: String,
    val link: String? = null,
    val tags: List<String>? = null,
    val language: String? = null,
    val images: List<String>? = null
)

class PostsDatabase(private val docs: DocumentsDatabase) {
    
    fun create(payload: PostPayload, targets: List<String>): Post {
        val payloadJson = json.encodeToString(PostPayload.serializer(), payload)
        val row = docs.createOrUpdate(
            kind = "post",
            payload = payloadJson,
            tags = targets.map { Tag(it, "target") }
        )
        return Post(
            uuid = row.uuid,
            createdAt = row.createdAt,
            targets = targets,
            content = payload.content,
            link = payload.link,
            tags = payload.tags,
            language = payload.language,
            images = payload.images
        )
    }
    
    fun getAll(): List<Post> {
        val rows = docs.getAll("post", "created_at DESC")
        return rows.map { row ->
            val payload = json.decodeFromString<PostPayload>(row.payload)
            Post(
                uuid = row.uuid,
                createdAt = row.createdAt,
                targets = row.tags.filter { it.kind == "target" }.map { it.name },
                content = payload.content,
                link = payload.link,
                tags = payload.tags,
                language = payload.language,
                images = payload.images
            )
        }
    }
    
    fun searchByUuid(uuid: String): Post? {
        val row = docs.searchByUuid(uuid) ?: return null
        val payload = json.decodeFromString<PostPayload>(row.payload)
        return Post(
            uuid = row.uuid,
            createdAt = row.createdAt,
            targets = row.tags.filter { it.kind == "target" }.map { it.name },
            content = payload.content,
            link = payload.link,
            tags = payload.tags,
            language = payload.language,
            images = payload.images
        )
    }
}

File: backend-scala/src/main/scala/socialpublish/db/PostsDatabase.scala
package socialpublish.db

import cats.effect.IO
import cats.syntax.all.*
import io.circe.parser.*
import io.circe.syntax.*
import socialpublish.models.*
import java.util.UUID

trait PostsDatabase:
  def create(content: String, link: Option[String], tags: List[String], 
             language: Option[String], images: List[UUID], targets: List[Target]): IO[Post]
  def getAll: IO[List[Post]]
  def searchByUUID(uuid: UUID): IO[Option[Post]]

class PostsDatabaseImpl(docsDb: DocumentsDatabase) extends PostsDatabase:
  
  override def create(
    content: String, 
    link: Option[String], 
    tags: List[String],
    language: Option[String],
    images: List[UUID],
    targets: List[Target]
  ): IO[Post] =
    import io.circe.generic.auto.*
    
    val payload = Map(
      "content" -> content.asJson,
      "link" -> link.asJson,
      "tags" -> tags.asJson,
      "language" -> language.asJson,
      "images" -> images.map(_.toString).asJson
    ).asJson.noSpaces
    
    val docTags = targets.map(t => DocumentTag(t.toString.toLowerCase, "target"))
    
    docsDb.createOrUpdate("post", payload, docTags).map { doc =>
      Post(
        uuid = doc.uuid,
        content = content,
        link = link,
        tags = tags,
        language = language,
        images = images,
        targets = targets,
        createdAt = doc.createdAt
      )
    }
  
  override def getAll: IO[List[Post]] =
    docsDb.getAll("post", "created_at DESC").flatMap { docs =>
      docs.traverse(docToPost)
    }
  
  override def searchByUUID(uuid: UUID): IO[Option[Post]] =
    docsDb.searchByUUID(uuid).flatMap {
      case Some(doc) => docToPost(doc).map(Some(_))
      case None => IO.pure(None)
    }
  
  private def docToPost(doc: Document): IO[Post] =
    import io.circe.generic.auto.*
    
    IO.fromEither(parse(doc.payload)).flatMap { json =>
      IO.fromEither(
        for
          content <- json.hcursor.get[String]("content")
          link <- json.hcursor.get[Option[String]]("link")
          tags <- json.hcursor.get[Option[List[String]]]("tags")
          language <- json.hcursor.get[Option[String]]("language")
          imagesStr <- json.hcursor.get[Option[List[String]]]("images")
        yield
          val images = imagesStr.getOrElse(Nil).map(UUID.fromString)
          val targets = doc.tags
            .filter(_.kind == "target")
            .map(t => Target.valueOf(t.name.capitalize))
          
          Post(
            uuid = doc.uuid,
            content = content,
            link = link,
            tags = tags.getOrElse(Nil),
            language = language,
            images = images,
            targets = targets,
            createdAt = doc.createdAt
          )
      )
    }
